### ISSUE TITLE
IterativeSolvers not harnessing multi-core machines?

### ISSUE THREAD

**Issue by ma-sadeghi** [JuliaMath/IterativeSolvers.jl#203]
Hi everyone,

I wonder if `IterativeSolvers` supports parallel execution in multicore machines. I checked my CPU usage while using `IterativeSolvers.cg!` for solving a 27-million-unknown sparse system and it seems that it's only using one of the cores.

I'd really appreciate your help.

**Comment by haampie** [comment]
Conjugate Gradients is pretty much a sequential iterative method, where each step consists of: 1 matrix-vector product, 2 dot products, 3 scalar times vector plus vector ops.

The matrix-vector product is the most consuming part and is at the same time not implemented by IterativeSolvers.jl. Is your matrix of type SparseMatrixCSC? At any rate, it might help to look into doing a threaded matrix-vector product. (Unfortunately threaded matrix-vector products with column-major matrices such as SparseMatrixCSC are pretty difficult to write, a row-major version lends itself better for the job)

Other than that: the operations are pretty much BLAS-1 (one operation per vector element), and threading typically does not help much in these cases. Also, the dot-products are done via OpenBLAS; if they're not threaded right now, they should be in a next Julia version, since OpenBLAS supports this only since not too long.

**Comment by ma-sadeghi** [comment]
@haampie Thanks a lot for your prompt reply. Yes, I'm using Julia's native `SparseMatrixCSC'. Can I ask what would you recommend for solving a large symmetric positive-definite sparse system in parallel?

**Comment by mschauer** [comment]
> Unfortunately threaded matrix-vector products with column-major matrices such as SparseMatrixCSC are pretty difficult to write

@haampie Wouldn't it be enough to circumvent the problem by writing x'A and using the transpose of a SparseMatrixCSC in the problem?

**Comment by andreasnoack** [comment]
The easiest solution here is to install MKL and load `https://github.com/JuliaSparse/MKLSparse.jl` since it will give you multithreaded sparse matvecs without doing anything else. That should give a significant effect on large problems.

**Comment by haampie** [comment]
@mschauer Right, A is symmetric, that makes life easier ^^.

Here's a simple native Julia approach, I'm curious if it helps a bit:

```julia
import Base: eltype, size
import Base.LinAlg: A_mul_B!
using Base.Threads

struct ThreadedMul{Tv,Ti}
    A::SparseMatrixCSC{Tv,Ti}
end
function A_mul_B!(y::AbstractVector, M::ThreadedMul, x::AbstractVector)
    Threads.@threads for i = 1 : M.A.n
        _threaded_mul!(y, M.A, x, i)
    end
    y
end
@inline function _threaded_mul!(y, A::SparseMatrixCSC{Tv}, x, i) where {Tv}
    s = zero(Tv)
    @inbounds for j = A.colptr[i] : A.colptr[i + 1] - 1
        s += A.nzval[j] * x[A.rowval[j]]
    end
    @inbounds y[i] = s
    y
end
eltype(M::ThreadedMul) = eltype(M.A)
size(M::ThreadedMul, I...) = size(M.A, I...)
```

```julia
using Base.Test

function random_spd_matrix(n::Int)
    A = sprand(n, n, 8 / n)
    A + A' + 8I
end

function test_threaded_mul(n = 100)
    A = random_spd_matrix(n)
    M = ThreadedMul(A)
    x, y = rand(n), rand(n)
    @test A_mul_B!(y, M, x) ≈ A * x
end
```

```julia
using BenchmarkTools
using IterativeSolvers

function bench_cg(n = 10_000)
    A = random_spd_matrix(n)
    M = ThreadedMul(A)
    b = rand(n)

    x_, hist = cg(A, b, maxiter = 50, log = true, tol = 1e-6)
    @show hist

    ours = @benchmark cg!(x, $M, $b, maxiter = 50, tol = 1e-6) setup = (x = ones($n))
    theirs = @benchmark cg!(x, $A, $b, maxiter = 50, tol = 1e-6) setup = (x = ones($n))

    ours, theirs
end
```

On my MacBook Air with just two cores I get something like:

```julia
julia> test_threaded_mul(100_000)
Test Passed

julia> a, b = bench_cg(10_000)
hist = Converged after 13 iterations.
(Trial(4.793 ms), Trial(6.043 ms))

julia> a, b = bench_cg(100_000)
hist = Converged after 13 iterations.
(Trial(84.728 ms), Trial(112.268 ms))
```

which looks OK. But then again, results depend on your matrix.

**Comment by ma-sadeghi** [comment]
Thanks @haampie, I tried your code but I get this error
```
ERROR: MethodError: no method matching *(::ThreadedMul{Float64,Int64}, ::Array{Float64,1})
Closest candidates are:
  *(::Any, ::Any, ::Any, ::Any...) at operators.jl:466
```

**Comment by haampie** [comment]
I did not implement `*`, only `A_mul_B!`. Something like
```
import Base: *
*(A::ThreadedMul, y::AbstractVector) = A_mul_B!(similar(y), A, y)
```
maybe helps.

**Comment by ma-sadeghi** [comment]
@haampie Thanks! Your solution worked like a charm! Just a quick question, I ran your code snippet (`bench_cg`) with `n=1_000_000` on a Linux machine with 8 cores (with `JULIA_NUM_THREADS=8`), and it scaled beautifully to almost 8x speed up. However, when I tried it on my custom matrix of roughly the same size, I can only get up to 2.5-3x speed up. Any ideas on the probably underlying reason? My matrix is a Laplacian in 3d, i.e. it has 7 diagonals on average.
